demos/basic/actors/zmqActor.py

```diff
@@ -0,0 +1,172 @@
+from improv.actor import Actor
+
+import zmq
+from zmq import PUB, SUB, SUBSCRIBE, REQ, REP, LINGER, Again
+from zmq.log.handlers import PUBHandler
+import traceback
+
+import zmq.asyncio
+import asyncio
+import time
+
+import logging
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class ZmqPSActor(Actor):
+    """
+    Zmq actor with PUB/SUB pattern.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.send_socket = None
+        self.recv_socket = None
+        self.address = None
+        self.context = None
+
+    def setSendSocket(self, ip, port, timeout=0.001):
+        """
+        Sets up the send socket for the actor.
+        """
+
+        self.context = zmq.Context()
+        self.send_socket = self.context.socket(PUB)
+        # bind to the socket according to the ip and port
+        self.address = "tcp://{}:{}".format(ip, port)
+        self.send_socket.bind(self.address)
+        time.sleep(timeout)
+
+    def setRecvSocket(self, ip, port, timeout=0.001):
+        """
+        Sets up the receive socket for the actor.
+        """
+
+        self.context = zmq.Context()
+        self.recv_socket = self.context.socket(SUB)
+        self.address = "tcp://{}:{}".format(ip, port)
+        self.recv_socket.connect(self.address)
+        self.recv_socket.setsockopt(SUBSCRIBE, b"")
+        time.sleep(timeout)
+
+    def sendMsg(self, msg):
+        """
+        Sends a message to the controller.
+        """
+
+        self.send_socket.send_pyobj(msg)
+        self.send_socket.close()
+        self.context.term()
+
+    def recvMsg(self):
+        """
+        Receives a message from the controller.
+        """
+
+        recv_msg = ""
+        while True:
+            try:
+                recv_msg = self.recv_socket.recv_pyobj(flags=zmq.NOBLOCK)
+                break
+            except Again:
+                pass
+        self.recv_socket.close()
+        self.context.term()
+        return recv_msg
+
+
+class ZmqRRActor(Actor):
+    """
+    Zmq actor with REQ/REP pattern.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.req_socket = None
+        self.rep_socket = None
+        self.address = None
+        self.context = None
+
+    def setReqSocket(self, ip, port, timeout=0.001):
+        """
+        Sets up the request socket for the actor.
+        """
+
+        self.context = zmq.asyncio.Context()
```

Comment:
To me, it seems more reasonable to have a single `Context` for the object that gets set up in the constructor.

Reply:
I found that if I change this to a single `Context` it will generate TypeError, so maybe we do need `asyncio` here

Reply:
I wouldn't think that's the problem unless you're mixing and matching. For a given method (blocking or asyncio, you should just be able to make a single context in the constructor and reuse `self.context`. Maybe send me the traceback on Slack so we can take a look together?

Reply:
I found that in tui.py, it also uses zmq.asyncio.Context() when setting up REQ socket. Though the original code in tui.py is that:
 ```python
self.context = zmq.Context()
self.control_socket = self.context.socket(REQ)
```
This is because that it imports zmq.asyncio as zmq:
```python
import zmq.asyncio as zmq
from zmq import PUB, SUB, SUBSCRIBE, REQ, REP, LINGER
```
Currently, I also changed `import zmq` to `import zmq.asyncio as zmq`, so that I can use a single Context in the code.

Reply:
I just found that if both version of zmqActor uses zmq.asyncio, then I cannot use zmq.NOBLOCK for PUB/SUB pattern, and it will generate an error:
```
 AssertionError: assert <Future pending cb=[_AsyncSocket._deserialize.<locals>._chain_cancel() at /opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/zmq/_future.py:364]> == 'hello'
```
since the type of the recvmsg now becomes future object. So I think for PUB/SUB, it requires zmq.Context() while REQ/REP requires zmq.asyncio.Context().

Reply:
Sorry, I don't fully understand how to achieve this. For the REQ/REP pattern, I used the send/receive example from the zmq documentation and it uses zmq.asyncio.Context() according to the documentation. For the PUB/SUB pattern, I failed to find an example that uses zmq.asyncio.Context(), can you give me some advice on that?

Reply:
I found that if I use a single Context() in the constructor like follows:
```python
 def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.send_socket = None
        self.recv_socket = None
        self.address = None
        self.context = zmq.Context()
```
Then the zmq demo cannot function normally. It will generate the following error:
```
actors.sample_generator --------------------------------Generator Exception: Operation not supported
actors.sample_generator Traceback (most recent call last):
  File "/Users/marleneli/draelos-lab/improv-proj/improv/demos/zmq_ps/actors/sample_generator.py", line 65, in runStep
    self.publish.setSendSocket(ip="127.0.0.1", port=5556)
  File "/Users/marleneli/draelos-lab/improv-proj/improv/demos/basic/actors/zmqActor.py", line 35, in setSendSocket
    self.send_socket = self.context.socket(PUB)
  File "/Users/marleneli/anaconda3/lib/python3.10/site-packages/zmq/sugar/context.py", line 359, in socket
    raise ZMQError(Errno.ENOTSUP)
zmq.error.ZMQError: Operation not supported
```

Reply:
I think this is because I need to term the Context each time after I finish one message transition.